### PR TITLE
Switch to epd_waveshare crate instead of il3820

### PR DESCRIPTION
**Pull request by caemor**
I think I got everything working now, I only might need to test the switch from partial refresh to full refresh to partial refresh a bit more and clean up a few comments

### CODE REVIEW

**Comment by TeXitoi** [comment]
Great! I'll try to review and test shortly!

You need to rust-fmt to make travis happy.

Thanks!

**Comment by caemor** [comment]
It let it run since the morning today and when I came back there was once more the problem that it doesn`t change back to the quick refresh lut after doing the full slow refresh at the full hour.

Commenting out the if-braces (https://github.com/caemor/rusty-clock/blob/master/src/main.rs#L285) helps so I might need to revert the changes from https://github.com/TeXitoi/rusty-clock/pull/18/commits/3df16098a6dfaad073fdf135c1d025278b9609d5 but then the LUT is set each second which is just uneccessary.

I am still unsure why it doesn`t work with the if-case.

**Comment by TeXitoi** [comment]
Maybe that's because sometime the command is sent while the screen ils busy. I've seen that you don't wait for busy to be stoped before sending commands in your driver.

**Comment by caemor** [comment]
Yeah I only used them for sleep and init at the moment. It seems I didn`t test them extensively enough when removing/reducing all the delays.  The delay after calling DISPLAY.display_frame is necessary and also reasonable since it doesn`t just need to write some data but rather needs some time to run all the Display Refreshes.

With the new fix it works fine, but I also released a patch for the epd_waveshare crate with the fixes for epd1in54 and epd2in9.

File: portable/src/ui/mod.rs [hunk 1]
```diff
@@ -148,8 +150,10 @@ impl Model {
         }
         cmds
     }
-    pub fn view(&self) -> DisplayRibbonLeft {
-        let mut display = DisplayRibbonLeft::default();
+    pub fn view(&self) -> Display2in9 {
+        let mut display = Display2in9::default();
+        display.set_rotation(DisplayRotation::Rotate270);
+        //let mut display = DisplayRibbonLeft::default();
```
Comment (TeXitoi) [inline portable/src/ui/mod.rs]: useless comment

File: src/main.rs [hunk 2]
```diff
@@ -266,14 +269,27 @@ const APP: () = {
             .FULL_UPDATE
             .lock(|fu| core::mem::replace(&mut *fu, false));
         if full_update {
-            resources.DISPLAY.set_full();
+            resources
+                .DISPLAY
+                .set_lut(&mut *resources.SPI, Some(RefreshLUT::FULL))
+                .unwrap(); //resources.DISPLAY.set_full();
```
Comment (TeXitoi) [inline src/main.rs]: useless command

File: src/main.rs [hunk 3]
```diff
@@ -266,14 +269,27 @@ const APP: () = {
...
-            .set_display(&mut *resources.SPI, &display)
+            .update_frame(&mut *resources.SPI, &display.buffer())
+            .unwrap();
+        resources
+            .DISPLAY
+            .display_frame(&mut *resources.SPI)
             .unwrap();
-        resources.DISPLAY.update(&mut *resources.SPI).unwrap();
-        resources.DISPLAY.set_partial();
+        while resources.DISPLAY.is_busy() {}
```
Comment (TeXitoi) [inline src/main.rs]: The busy wait should be in the if, allowing the microcontroller to sleep during the rendering

File: src/main.rs [hunk 4]
```diff
@@ -266,14 +269,27 @@ const APP: () = {
...
+        if full_update {
+            // partial/quick refresh needs only be set when a full update was run before
+            resources
+                .DISPLAY
+                .set_lut(&mut *resources.SPI, Some(RefreshLUT::QUICK))
+                .unwrap();
+        } //replacement of old: resources.DISPLAY.set_partial();
```
Comment (TeXitoi) [inline src/main.rs]: useless comment

**Review by TeXitoi — COMMENTED** [review]
First, sorry to only came back now.

I'm a bit disappointed by the manual busy wait: for me, this busy wait must be done before each command send, and thus enforced in the driver directly.

**Comment by caemor** [comment]
I removed all the old comments and added busy_waits everywhere directly in the new version of the epd_waveshare crate.

**Review by TeXitoi — APPROVED** [review]

**Comment by TeXitoi** [comment]
Great!

**Comment by TeXitoi** [comment]
Looking at https://github.com/caemor/epd-waveshare/commit/7b4a7f057855a43895c984aed5c2a542827a8828 I think it would be better if the wait is done before the command, not after. This way, you can sleep while some work is done on the screen, and wait to the screen being ready before sending another command.

In our case, it would allow to go to sleep during the refresh of the screen, but waiting for the screen to be ready before sending a new image.

**Comment by caemor** [comment]
That makes sense. I'm going to change this for the next release then.

**Comment by TeXitoi** [comment]
I'll try to flash it tonight before merging.

**Comment by TeXitoi** [comment]
Works as expected. Thanks!

**Pull request merged**